dMagnetic.1
.\" Process this file with
.\" groff -man -Tascii dMagnetic.1
.\"
.Dd December 24th,2020
.Os OpenBSD 
.Dt dMagnetic 1
.
.Sh NAME
.Nm dMagnetic
.Nd A Magnetic Scrolls Interpreter
.
.
.Sh INTRODUCTION
You wake up on a sunny August morning with birds singing, and the air fresh and clear. However, your joints are stiff and you have not woken up in your bedroom as you would have expected. Trying to recall what happened the night before, you manage to piece together a few brief glimpses to give the following account...
.Sh SYNOPSIS
.
.Nm
.Op Fl mag Ar MAGFILE.mag
.Op Fl gfx Ar GFXFILE.gfx
.br
.Nm
.Op Fl msdosdir Ar DIRECTORY/
.Op Fl vmode Ar high_ansi2
.br
.Nm
.Op Fl tworsc Ar DIRECTORY/TWO.RSC
.Op Fl vmode Ar high_ansi
.br
.Nm
.Op Fl d64 Ar DIRECTORY/FILE1.d64,DIRECTORY/FILE2.d64
.Op Fl vmode Ar sixel
.br
.Nm
.Op Fl amstradcpc Ar DIRECTORY/FILE1.DSK,DIRECTORY/FILE2.DSK
.Op Fl vmode Ar utf
.br
.Nm
.Op Fl spectrum Ar DIRECTORY/DISKIMAGE.DSK
.Op Fl vmode Ar none
.br
.Nm
.Op Fl archimedes Ar DIRECTORY/DISKIMAGE.adf
.br
.Nm
.Op Fl atarixl Ar DIRECTORY/FILE1.ATR,DIRECTORY/FILE2.ATR
.br
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.br
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vrows Ar ROWS
.Op Fl vcols Ar COLUMNS
.br
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vmode Ar none
.br
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vmode Ar monochrome
.br
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vmode Ar low_ansi
.br
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vmode Ar low_ansi2
.br
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vmode Ar high_ansi
.br
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vmode Ar high_ansi2
.br
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vmode Ar utf
.br
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vmode Ar sixel
.Op Fl sres Ar 1024x768
.br
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl valign Ar left
.br
.Nm
.Op Fl vlog Ar LOGFILE 
.Op Fl vecho
.Nm
.Op Fl help
.Op Fl helpini
.Op Fl v
.Op Fl bsd
.Sh DESCRIPTION
.Nm
is an interpreter for classic text adventure games, or interactive fiction, such as "The Pawn" and "The Guild of Thieves". The following games are being supported:
.Bl -tag -width 10
.It "The Pawn"
.It "The Guild of Thieves"
.It "Jinxter"
.It "Fish!"
.It "Myth"
.It "Corruption"
.It "Wonderland"
.El
.
.Sh GAME BINARIES
Currently, binaries in the .mag and .gfx format from the 
magnetic scrolls memorial at
https://msmemorial.if-legends.org/memorial.php 
are being supported. 
Once downloaded, their position has to be made available to 
.Nm
by editing dMagnetic.ini first. 
.br
Other possibilities are loading the games from original releases, such as i
the MS-DOS binaries, 
the Magnetic Windows Resource Files, 
the Commodore C64 images in the d64 format, 
Amstrad CPC and Spectrum128 images in the DSK format.
Acorn Archimedes images in the ADF/ADL (aka ADFS) format.
and finally
AtariXL/Atari800 images in the ATR format.
.Sh STARTING THE GAME
Once the game(s) have been downloaded, the .mag and .gfx files can be provided via
.Pp
.Nm
-mag pawn.mag -gfx pawn.gfx
.Pp
Alternatively, binaries from the MSDOS version of 
"The Pawn"
or
"The Guild of Thieves"
can be used. The directory, in which they are, has to be provided via
.Pp
.Nm
-msdosdir DIRECTORY/ -vmode high_ansi
.Pp 
It is recommended to use the 
.Op Fl vmode Ar high_ansi2
option, since this renders the half-tone images slightly different.
.Pp
A third option is to use the resource files from "Wonderland" or the magnetic scrolls collection. For this, the location of the file ending with "TWO.RSC" has to be provided.
.Pp
.Nm
-tworsc DIRECTORY/TWO.RSC -vmode high_ansi
.br
.Nm
-tworsc DIRECTORY/GTWO.RSC -vmode high_ansi -ega
.br
The names of the other resource files is being calculated from that. With the optional parameter -ega, the EGA versions of the pictures are being shown.
.Pp
A fourth option is to use D64 image files from the Commodore 64 releases. They come in pairs, since noth sides of the floppy disk were occupied. So the parameter requires a list of them, separated by komma:
.Pp
.Nm
-d64 DIRECTORY/FILE1.d64,DIRECTORY/FILE2.d64 -vmode sixel
.Pp
A fifth option is to use DSK image files from the Amstrad CPC releases. They come in pairs, since noth sides of the floppy disk were occupied. So the parameter requires a list of them, separated by komma:
.Pp
.Nm
-amstradcpc DIRECTORY/FILE1.DSK,DIRECTORY/FILE2.DSK -vmode sixel
.Pp
A sixth option is to use DSK image files from the Spectrum releases. Since this platform did not have the pictures, a single floppy disk was enough:
.Pp
.Nm
-spectrum DIRECTORY/DISKIMAGE.DSK -vmode none
.Pp
A seventh option is to use ADF/ADL/ADFS image files from the Acorn Archimedes releases. 
.Pp
.Nm
-archimedes DIRECTORY/DISKIMAGE.ADF 
.Pp
An eigth option is to use ATR image files from the AtariXL/Atari800 releases. Those come in pairs.
.Pp
.Nm
-atarixl DIRECTORY/FILE1.ATR,DIRECTORY/FILE2.ATR
.Pp
It is also possible to edit the dMagnetic.ini file and start a specific game, by using one of
.Pp
.Nm
-ini dMagnetic.ini pawn
.br
.Nm
-ini dMagnetic.ini guild
.br
.Nm
-ini dMagnetic.ini jinxter
.br
.Nm
-ini dMagnetic.ini corruption
.br
.Nm
-ini dMagnetic.ini fish
.br
.Nm
-ini dMagnetic.ini myth
.br
.Nm
-ini dMagnetic.ini wonderland
.Pp
When you see the prompt, the game is ready to accept your commands. For example EXAMINE CLOTHES. Or GO EAST. Or ASK KRONOS ABOUT THE WRISTBAND.
.br
.Sh LEAVING THE GAME
Type in "QUIT". Duh!
.Sh SAVING/LOADING PROGRESS
The game can be saved at any type by typing SAVE, and loaded by typing LOAD. Followed by a filename.
.Pp
SAVE myprogress1.sav
.br
LOAD myprogress1.sav
.Sh TEXT ALIGNMENT
Changing the alignment of the output text can be done with the .ini file or the commandline.
.br
.Nm Fl valign Ar left
.br
.Nm Fl valign Ar block
.br
.Nm Fl valign Ar right
.br
.Sh GRAPHIC MODES
.Nm
has been developed with ANSI-consoles in mind. To select a video output that best suits your needs, please try one of the following
.
.Pp
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vmode Ar none
.br
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vmode Ar monochrome
.br
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vmode Ar low_ansi
.br
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vmode Ar low_ansi2
.br
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vmode Ar high_ansi
.br
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vmode Ar high_ansi2
.br
.Pp
.Nm
.Op Fl ini Ar ini-file
.Op GAME
.Op Fl vmode Ar utf
.Pp
.br
To change the number of rows/columns to render the images, use
.Pp
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vcols Ar COLUMNS
.br
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vrows Ar ROWS
.Pp
Sixels are a signalling mode for some terminal emulators (like mlterm, or xterm -ti vt340
, for example), that offer the possibility of drawing high resolution images. It can be selected with
.Pp
.Nm
.Op Fl ini Ar ini-file 
.Op GAME
.Op Fl vmode Ar sixel
.Op Fl sres Ar 1024x768
.Pp
where the -sres 
option offers a way to scale the images. The maximum allowed width is 1024 pixels.
.Sh GRAPHICS IN WONDERLAND AND THE MAGNETIC SCROLLS COLLECTION
Before you can see the beautiful graphics in "Wonderland", or from the Magnetic Scrolls Collection, you have to type in "graphics" to enable them.
.Sh COMMAND LINE OPTIONS
.Op Fl bsd
.br
Shows the license.
.Pp
.Op Fl Fl help
.br
Shows the detailed help.
.Pp
.Op Fl Fl helpini
.br
Shows an example for a working dMagnetic.ini file.
.Pp
.Op Fl mag Ar MAGFILE.mag
.Op Fl gfx Ar GFXFILE.gfx
.br
.Pp
.Op Fl msdosdir Ar DIRECTORY/
.br
.Pp
.Op Fl tworsc Ar DIRECTORY/TWO.RSC
.br
.Pp
.Op Fl d64 Ar DIRECTORY/FILE1.d64,DIRECTORY/FILE2.d64
.br
.Pp
.Op Fl amstradcpc Ar DIRECTORY/FILE1.DSK,DIRECTORY/FILE2.DSK
.br
.Pp
.Op Fl spectrum Ar DIRECTORY/DISKIMAGES.DSK
.br
.Pp
.Op Fl archimedes Ar DIRECTORY/DISKIMAGES.ADF
.Pp
.br
.Op Fl atarixl Ar DIRECTORY/FILE1.ATR,DIRECTORY/FILE2.ATR
.br
.Nm
is a Magnetic Scrolls Interpreter. To actually play the games, their binaries have to be provided. Either in the .mag and .gfx format from https://msmemorial.if-legends.org/magnetic.php, as the name of the directory in which the original MS-DOS version can be found, the location and the name of the second resource file TWO.RSC, as D64-images from the Commodore 64 releases, 
as DSK-images from the Amstrad/Schneider CPC releases,
as DSK-images from the Spectrum releases,
as ADF/ADL/ADFS images from the Acorn Archimedes releases,
or even as ATR images from the AtariXL/Atari800 releases.
.Pp
.Op Fl ini Ar dMagnetic.ini pawn
.br
.Op Fl ini Ar dMagnetic.ini guild
.br
.Op Fl ini Ar dMagnetic.ini jinxter
.br
.Op Fl ini Ar dMagnetic.ini corruption
.br
.Op Fl ini Ar dMagnetic.ini fish
.br
.Op Fl ini Ar dMagnetic.ini pawn
.br
.Op Fl ini Ar dMagnetic.ini myth
.br
.Op Fl ini Ar dMagnetic.ini wonderland
.br
An alternative way to provide the location of the .mag and .gfx, the image files or the directory name, is through a dMagnetic.ini file. See helpini or
.Xr dMagneticini 5
for an example of a working dMagnetic.ini file.
.br
IT SHOULD BE NOTED that the default location for dMagnetic.ini is in the user's home directory. If the file is located there, "The Pawn", for example, can be started by typing
.br
.Nm
pawn
.br
.Pp
.Op Fl rmode Ar pseudo
.Op Fl rseed Ar SEED
.br
.Op Fl rmode Ar real
.br
Certain elements of the game rely on chance. For this, the virtual machine within
.Nm
offers two possibilities: Playing against a "pseudo" random generator, which results in a certain degree of determinism. On the other hand, playing with "real" random values from the operating system, allows for a completely new experience. 
.Pp
.Op Fl vrows Ar ROWS
.br
.Op Fl vcols Ar COLUMNS
.br
.Nm
has been designed with terminal windows as main output in mind. The terminal window has a fixed number of rows and columns to print out text. Its size is limited, but it is used by 
.Nm
to render the beautiful BEAUTIFUL pictures as well. To restrict the space in which they are rendered (in glorious ANSI art), those command set the upper limits.
.br
.Pp
.Op Fl vecho
.br
When trying to run
.Nm
and redirecting the output into a file, the inputs are missing. This option reprints what was typed in, to allow for a spoilery script.
.Pp
.Op Fl vlog Ar LOGFILE.log
.br
Sort of a travel journal, this option lets you write the commands that where typed into a file.
.br
.Pp
.Op Fl vmode Ar none
.br
.Op Fl vmode Ar monochrome
.br
.Op Fl vmode Ar low_ansi
.br
.Op Fl vmode Ar low_ansi2
.br
.Op Fl vmode Ar high_ansi
.br
.Op Fl vmode Ar high_ansi2
.br
.Op Fl vmode Ar utf
.br
.Op Fl vmode Ar sixel
.Op Fl sres Ar 1024x768
.br
This option allows for selecting a different mode to render the images, should the actual one prove to be unsuitable for the preferred terminal program.
.br
.Pp
.Op Fl version
.br
Shows the current version of
.Nm
.
.Sh BUGS
Report bugs to
.An Aq dettus@dettus.net .
Make sure to include DMAGNETIC somewhere in the subject.
.Sh AUTHOR
Written by
.An Thomas Dettbarn
.Sh SEE ALSO
.Xr dMagneticini 5


src/toplevel/dMagnetic.c
/*

Copyright 2022, dettus@dettus.net

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this 
   list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice, 
   this list of conditions and the following disclaimer in the documentation 
   and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED 
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE 
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL 
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, 
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "version.h"
#include "configuration.h"
#include "vm68k.h"
#include "linea.h"
#include "default_callbacks.h"
#include "maggfxloader.h"
#include "dMagnetic_helpscreens.h"
#include "dMagnetic_pathnames.h"
#define	SHAREDMEMSIZE	98304	// 98304 bytes ought to be enough for everybody
#define	MAXMAGSIZE	(1<<20)
#define	MAXGFXSIZE	(1<<22)


int dMagnetic_init(void** hVM68k,void** hLineA,void* pSharedMem,int memsize,char* magbuf,int magsize,char* gfxbuf,int gfxsize)
{
	int sizevm68k;
	int sizelineA;
	int retval;
	int version;




	// step 2: start the engine.
	retval=vm68k_getsize(&sizevm68k);
	if (retval) 
	{
		fprintf(stderr,"ERROR: vm68k_getsize returned %d\n",retval);
		return retval;
	}
	retval=lineA_getsize(&sizelineA);
	if (retval) 
	{
		fprintf(stderr,"ERROR: lineA_getsize returned %d\n",retval);
		return retval;
	}
	*hVM68k=malloc(sizevm68k);
	*hLineA=malloc(sizelineA);
	memset(*hVM68k,0,sizevm68k);
	memset(*hLineA,0,sizelineA);
	if (*hVM68k==NULL || *hLineA==NULL) 
	{
		fprintf(stderr,"ERROR: unable to allocate memory for the engine\n");
		return -1;
	}
	pSharedMem=vm68k_getpSharedMem(*hVM68k);	
	retval=lineA_init(*hLineA,pSharedMem,&memsize,magbuf,magsize,gfxbuf,gfxsize);
	if (retval)
	{
		fprintf(stderr,"ERROR: lineA_init returned %d\n",retval);
		return retval;
	}
	retval=lineA_getVersion(*hLineA,&version);
	if (retval)
	{
		fprintf(stderr,"ERROR: lineA_getversion returned %d\n",retval);
		return retval;
	}
	fprintf(stderr,"Initializing the 68K version %d VM\n",version);	
	if (version==4)
	{
		fprintf(stderr,"\n");
		fprintf(stderr,"---------------------------------------\n");
		fprintf(stderr,"- Version 4 of the VM requires you to -\n");
		fprintf(stderr,"- activate the graphics manually. Use -\n");
		fprintf(stderr,"- the command       GRAPHICS      now -\n");
		fprintf(stderr,"---------------------------------------\n");
		fprintf(stderr,"\n");
	}
	retval=vm68k_init(*hVM68k,version);
	if (retval)
	{
		fprintf(stderr,"ERROR: vm68k_init returned %d\n",retval);
		return retval;
	}
	return 0;
}
int main(int argc,char** argv)
{
	int i;
	char inifilename[1024];
	int retval;
	FILE *f_inifile=NULL;

	void* hVM68k;
	void* hLineA;
	void* hGUI;
	void* pSharedMem;
	int sizeGUI;
	int unknownopcode;
	char *homedir;
	char random_mode;
	unsigned int random_seed;
	int egamode;
	char* magbuf;	
	char* gfxbuf;
	


	// figure out the location of the inifile.
	if (!(retrievefromcommandline(argc,argv,"--version",NULL,0)))
	{
		dMagnetic_helpscreens_header();
#define	LOCNUM	16
		const char *locations[LOCNUM]={
			PATH_ETC,
			PATH_USR_LOCAL_SHARE,
			PATH_USR_LOCAL_SHARE_GAMES,
			PATH_USR_LOCAL_SHARE"dMagnetic/",
			PATH_USR_LOCAL_SHARE_GAMES"dMagnetic/",
			PATH_USR_LOCAL_GAMES,
			PATH_USR_LOCAL_GAMES"dMagnetic/",
			PATH_USR_SHARE,
			PATH_USR_SHARE_GAMES,
			PATH_USR_SHARE_GAMES"dMagnetic/",
			PATH_USR_SHARE"dMagnetic/",
			PATH_USR_GAMES,
			PATH_USR_GAMES"dMagnetic/",
			PATH_USR_SHARE"doc/dmagnetic/",
			PATH_USR_PKG_SHARE"doc/dMagnetic/",
			"./"};	// this should always be the last one.

		f_inifile=NULL;
		if (f_inifile==NULL)
		{
			homedir=getenv("HOME");
			snprintf(inifilename,1023,"%s/dMagnetic.ini",homedir);
			f_inifile=fopen(inifilename,"rb");
		}
		if (f_inifile==NULL)
		{
			homedir=getenv("HOME");
			snprintf(inifilename,1023,"%s/.dMagnetic.ini",homedir);
			f_inifile=fopen(inifilename,"rb");
		}
		for (i=0;i<LOCNUM;i++)
		{
			if (f_inifile==NULL)
			{
				snprintf(inifilename,1023,"%sdMagnetic.ini",locations[i]);
				f_inifile=fopen(inifilename,"rb");
			}
		}

		if (f_inifile) 
		{
			fclose(f_inifile);
		}
	}	
	if (argc<2)
	{
		dMagnetic_helpscreens_loaderfailed(argv[0]);
		return 1;
	}
	if ((retrievefromcommandline(argc,argv,"--version",NULL,0))
		|| (retrievefromcommandline(argc,argv,"-version",NULL,0))
		|| (retrievefromcommandline(argc,argv,"-v",NULL,0)))
	{
		printf("%d.%d%d\n",VERSION_MAJOR,VERSION_MINOR,VERSION_REVISION);
		return 0;
	}
	if (retrievefromcommandline(argc,argv,"-bsd",NULL,0))
	{
		dMagnetic_helpscreens_license();
		return 0;


	}
	if ((retrievefromcommandline(argc,argv,"--help",NULL,0))
	||  (retrievefromcommandline(argc,argv,"-help",NULL,0)))
	{
		dMagnetic_helpscreens_help(argv[0]);
		return 0;
	}
	if ((retrievefromcommandline(argc,argv,"--helpini",NULL,0))
	||  (retrievefromcommandline(argc,argv,"-helpini",NULL,0)))
	{
		dMagnetic_helpscreens_helpini();
		return 0;
	}
	if (retrievefromcommandline(argc,argv,"-ini",inifilename,sizeof(inifilename))) 
	{
	}
	fprintf(stderr,"Using .ini file: %s\n",inifilename);		
	f_inifile=fopen(inifilename,"rb");
	pSharedMem=malloc(SHAREDMEMSIZE);
	if (pSharedMem==NULL)
	{
		fprintf(stderr,"ERROR: unable to allocate shared memory\n");
		return 1;
	}
	retval=default_getsize(&sizeGUI);
	if (retval)
	{
		fprintf(stderr,"ERROR. default_getsize returned %d\n",retval);
		return 1;
	}
	hGUI=malloc(sizeGUI);
	if (hGUI==NULL)
	{
		fprintf(stderr,"ERROR: unable to locate memory for the GUI\n");
		return 1;
	}
	retval=default_open(hGUI,f_inifile,argc,argv);
	if (retval)
	{
		fprintf(stderr,"ERROR: opening the GUI failed\n");
		return 1;
	}

//////////////////////////////////////////////// random
	random_mode=0;
	random_seed=12345;
	if (f_inifile)
	{
		char result[64];

		if (retrievefromini(f_inifile,"[RANDOM]","mode",result,sizeof(result)))
		{
			if (result[0]=='p') random_mode=0;
			else if (result[0]=='r') random_mode=1;
			else {
				printf("illegal random mode in inifile. please use one of ");
				printf("pseudo ");
				printf("real ");
				return 1;
			}
		}
		if (retrievefromini(f_inifile,"[RANDOM]","seed",result,sizeof(result)))
		{
			random_seed=atoi(result);
			if (random_seed<1 || random_seed>0x7fffffff)
			{
				printf("illegal random seed. please use a value between %d and %d\n",1,0x7fffffff);
				return 1;
			}
		}
	}
		

	if (argc)
	{
		char result[64];
		if (retrievefromcommandline(argc,argv,"-rmode",result,sizeof(result)))
		{
			if (result[0]=='p') random_mode=0;
			else if (result[0]=='r') random_mode=1;
			else {
				printf("illegal parameter for -rmode. please use one of ");
				printf("pseudo ");
				printf("real ");
				printf("\n");
				return 1;
			}
		}
		if (retrievefromcommandline(argc,argv,"-rseed",result,sizeof(result)))
		{
			random_seed=atoi(result);
			if (random_seed<1 || random_seed>0x7fffffff)
			{
				printf("illegal parameter for -rseed. please use a value between %d and %d\n",1,0x7fffffff);
				return 1;
			}
		}
	}	
	egamode=0;
	if (f_inifile)
	{
		char result[64];
		if (retrievefromcommandline(argc,argv,"-ega",result,sizeof(result)))
		{
			egamode=1;
		}
	}


	if (f_inifile) fclose(f_inifile);
	magbuf=malloc(MAXMAGSIZE);
	gfxbuf=malloc(MAXGFXSIZE);

	// this is the main loop.
	do
	{
		int magsize;
		int gfxsize;

		magsize=MAXMAGSIZE;
		gfxsize=MAXGFXSIZE;
		if (magbuf==NULL || gfxbuf==NULL) 
		{
			fprintf(stderr,"ERROR: unable to allocate memory for the data files\n");
			return -1;
		}
		f_inifile=fopen(inifilename,"rb");
		if (loader_init(argc,argv,f_inifile, magbuf,&magsize,gfxbuf,&gfxsize))
		{
			dMagnetic_helpscreens_loaderfailed(argv[0]);
			return 1;
		}


		retval=dMagnetic_init(&hVM68k,&hLineA,pSharedMem,SHAREDMEMSIZE,magbuf,magsize,gfxbuf,gfxsize);
		if (retval)
		{
			return 1;
		}
		retval|=lineA_configrandom(hLineA,random_mode,random_seed);
		retval|=lineA_setEGAMode(hLineA,egamode);
		// set the call back hooks for this GUI
		retval|=lineA_setCBoutputChar(hLineA,default_cbOutputChar,	hGUI);
		retval|=lineA_setCBoutputString(hLineA,default_cbOutputString,	hGUI);
		retval|=lineA_setCBinputString(hLineA,default_cbInputString,	hGUI);
		retval|=lineA_setCBDrawPicture(hLineA,default_cbDrawPicture,	hGUI);
		retval|=lineA_setCBLoadGame(hLineA,default_cbLoadGame,		hGUI);
		retval|=lineA_setCBSaveGame(hLineA,default_cbSaveGame,		hGUI);
		if (retval)
		{
			fprintf(stderr,"ERROR: setting the API hooks failed\n");
			return 1;
		}
		/////////////////////////////////////////////		

		lineA_showTitleScreen(hLineA);	// some versions of the game have a title screen.
		// everything is good. have fun playing!
		do
		{
			unsigned short opcode;
			unknownopcode=0;
			retval=vm68k_getNextOpcode(hVM68k,&opcode);
			if (retval==VM68K_OK) retval=lineA_substitute_aliases(hLineA,&opcode);
			if (retval==LINEA_OK) retval=vm68k_singlestep(hVM68k,opcode);
			if (retval!=VM68K_OK) retval=lineA_singlestep(hLineA,hVM68k,opcode);
			if (retval!=LINEA_OK && retval!=LINEA_OK_QUIT && retval!=LINEA_OK_RESTART) 
			{
				fprintf(stderr,"\x1b[0m\n\x1b[0;37;44m Sorry. Unknown opcode %04X\x1b[0m\n",opcode);
				unknownopcode=1;
			}
			//		if (retval==LINEA_OK_QUIT) printf("\x1b[0m\n\x1b[0;37;44mGoodbye!\x1b[0m\n");
		} while (!unknownopcode && !retval);	
	} while (retval==LINEA_OK_RESTART);
	free(gfxbuf);
	free(magbuf);
	// this concludes the main loop

	free(pSharedMem);
	free(hGUI);
	free(hLineA);
	free(hVM68k);
	return 0;
}
